<img src="https://docs.google.com/uc?export=download&id=1EiHSYfHYk8nKMEWd6A74CMFVak5Lf4ab">

# Data manipulation: Data Files, and data cleaning and preparation


*  All the data used in this course, are downloaded from: [pydata-book](https://github.com/wesm/pydata-book.git) 
* You will not have to download any thing, just run the following cell:


In [1]:
!ls
!rm -R AAA-Ped-Week2
#downloading the needed data 
!git clone https://github.com/MostaSchoolOfAI/AAA-Ped-Week2.git
!ls
%cd AAA-Ped-Week2
import pandas as pd


AAA-Ped-Week2  db.sqlite  example2.db  example3.db  example.db	sample_data
Cloning into 'AAA-Ped-Week2'...
remote: Enumerating objects: 49, done.
remote: Counting objects: 100% (49/49), done.
remote: Compressing objects: 100% (41/41), done.
remote: Total 49 (delta 5), reused 36 (delta 2), pack-reused 0
Unpacking objects: 100% (49/49), done.
AAA-Ped-Week2  db.sqlite  example2.db  example3.db  example.db	sample_data
/content/AAA-Ped-Week2


# 1- Data Files: reading and writing

* Using **pandas**, we can easily **read** (and **write**) different **types** of **data** from: on **disk file**s,** interactions **with **web api,** or interactions with **databases**.

## On Disk Files

* You have just to **choose** the **right** function to use with the **right **parameters: 

In [7]:
#viewing the content of A3P-w2-ex1.csv
!cat A3P-w2-ex1.csv

a,b,c,d,message
1,2,3,4,hello
5,6,7,8,world
9,10,11,12,foo

In [50]:

# read the file
f1= pd.read_csv("A3P-w2-ex1.csv")
f1


a   b   c   d message
0  1   2   3   4   hello
1  5   6   7   8   world
2  9  10  11  12     foo

In [9]:
# some files can have no header, but not this one
# so the header is considered as a value (which is in this case wrong)
f2= pd.read_csv("A3P-w2-ex1.csv", header=None,sep=",")
f2

0   1   2   3        4
0  a   b   c   d  message
1  1   2   3   4    hello
2  5   6   7   8    world
3  9  10  11  12      foo

* For files with **no header**, a **default** columns **labels** are added
* But  you can add your own header  as **argument**
* You can alsor specify one of the  **columns** as an **index** column





In [18]:
# viewing the content of A3P-w2-ex2.csv
!cat A3P-w2-ex2.csv

1,2,3,4,hello
5,6,7,8,world
9,10,11,12,foo

In [24]:
# you can specify a header, and an index column
header=[ "col"+str(i) for i in range(1,6)]
ind= header[len(header)-1]
f3= pd.read_csv("A3P-w2-ex2.csv", names=header,index_col=ind)
f3

col1  col2  col3  col4
col5                         
hello     1     2     3     4
world     5     6     7     8
foo       9    10    11    12

* Other **arguments** exist, that allow us to
  *   Specify the values to consider as a** Null values**
  *   Specify the **numbers of rows** to be read
  *   Read the file in **multiple chunks**


In [10]:
!cat "A3P-w2-ex5.csv"

something,a,b,c,d,message
one,1,2,3,4,NA
two,5,6,,8,world
three,9,10,11,12,foo

In [11]:
pd.read_csv("A3P-w2-ex5.csv")
            

something  a   b     c   d message
0       one  1   2   3.0   4     NaN
1       two  5   6   NaN   8   world
2     three  9  10  11.0  12     foo

In [48]:
# considering also: the 5 value for "a" column, the 2 and 10 values for "b" column as null
pd.read_csv("A3P-w2-ex5.csv",na_values={"a":5,"b":[2,10]})

something    a    b     c   d message
0       one  1.0  2.0   3.0   4     NaN
1       two  5.0  6.0   NaN   8   world
2     three  NaN  NaN  11.0  12     foo

In [23]:
# the initial file: number of rows = 10000
df=pd.read_csv("A3P-w2-ex6.csv")
df.shape

(10000, 5)

In [46]:
# we will read only 10 rows, sepearated in 5 chunks of 2 rows
tfr=pd.read_csv("A3P-w2-ex6.csv",nrows =10, chunksize=2)

for i, chunk in zip(range(1,6),tfr):
  print("chunk"+str(i)+":\n", chunk)


chunk1:
         one       two     three      four key
0  0.467976 -0.038649 -0.295344 -1.824726   L
1 -0.358893  1.404453  0.704965 -0.200638   B
chunk2:
         one       two     three      four key
2 -0.501840  0.659254 -0.421691 -0.057688   G
3  0.204886  1.074134  1.388361 -0.982404   R
chunk3:
         one       two     three      four key
4  0.354628 -0.133116  0.283763 -0.837063   Q
5  1.817480  0.742273  0.419395 -2.251035   Q
chunk4:
         one       two     three      four key
6 -0.776764  0.935518 -0.332872 -1.875641   U
7 -0.913135  1.530624 -0.572657  0.477252   K
chunk5:
         one       two     three      four key
8  0.358480 -0.497572 -0.367016  0.507702   S
9 -1.740877 -1.160417 -1.637830  2.172201   G


* With **read_csv** or **read_table**, you can read other text files format as (**.txt** files) containing columns separated by delimiters.
* You can use** read_json** to read** json** files 
* You can use **read_html** to read tabular data in a **html** file.
* You can use** read_hdf** to read the binary **hdf5** format, or **read_excel **to read **excel **files.

In [58]:
! cat A3P-w2-example.json	

[{"a": 1, "b": 2, "c": 3},
 {"a": 4, "b": 5, "c": 6},
 {"a": 7, "b": 8, "c": 9}]


In [60]:
pd.read_json("A3P-w2-example.json"	)

a  b  c
0  1  2  3
1  4  5  6
2  7  8  9

In [10]:
# here how the html file looks like
# it has one table
%%html
A3P-w2-fdic_failed_bank_list.html

In [4]:
# we have to install lxml, 
!pip install lxml 
# we have to install  beautifulsoup4 and html5lib too (if they are not installed)

!pip install beautifulsoup4

!pip install html5lib
# you will have to restart the runtime ( an run again the firs cell)


    100% |████████████████████████████████| 5.8MB 5.0MB/s 


In [3]:
# we have to install  beautifulsoup4 and html5lib too (if they are not installed)
# you will have to restart the runtime ( an run again the firs cell)

!pip install beautifulsoup4

!pip install html5lib

    100% |████████████████████████████████| 122kB 5.8MB/s 


In [2]:

dfs= pd.read_html("A3P-w2-fdic_failed_bank_list.html")
print("number of tables ==", len(dfs))
# only the 5  rows will be displayed by default
pd.options.display.max_rows = 5
dfs[0]

number of tables == 1


Bank Name        City  ST   CERT  \
0                          Allied Bank    Mulberry  AR     91   
1         The Woodbury Banking Company    Woodbury  GA  11297   
..                                 ...         ...  ..    ...   
545  National State Bank of Metropolis  Metropolis  IL   3815   
546                   Bank of Honolulu    Honolulu  HI  21029   

       Acquiring Institution        Closing Date       Updated Date  
0               Today's Bank  September 23, 2016  November 17, 2016  
1                United Bank     August 19, 2016  November 17, 2016  
..                       ...                 ...                ...  
545  Banterra Bank of Marion   December 14, 2000     March 17, 2005  
546       Bank of the Orient    October 13, 2000     March 17, 2005  

[547 rows x 7 columns]

In [3]:
# to read excel files using pandas, you have first to install this packages
!pip install xlrd 
!pip install openpyxl

    100% |████████████████████████████████| 112kB 2.5MB/s 
    100% |████████████████████████████████| 2.0MB 7.2MB/s 
  Running setup.py bdist_wheel for openpyxl ... - \ | done
  Stored in directory: /root/.cache/pip/wheels/3f/37/28/5ab3dffb7ff261e6fa21455ec9d157f95958e818c6b89f024c
  Running setup.py bdist_wheel for et-xmlfile ... - done
  Stored in directory: /root/.cache/pip/wheels/2a/77/35/0da0965a057698121fc7d8c5a7a9955cdbfb3cc4e2423cad39
Successfully built openpyxl et-xmlfile


In [4]:
pd.read_excel("A3P-w2-ex1.xlsx", 'Sheet1')

a   b   c   d message
0  1   2   3   4   hello
1  5   6   7   8   world
2  9  10  11  12     foo

* To **write** the data to a file, you can use this **corresponding**
 methods: to_csv, to_json, and to excel.
 

In [19]:
df = pd.DataFrame([range(1,4),range(5,8),range(7,4,-1)],index=range(1,4),columns=list("abc"))
df

a  b  c
1  1  2  3
2  5  6  7
3  7  6  5

In [0]:
df.to_csv("file1.csv")
df.to_csv("file1.txt")
df.to_json("file1.json")
df.to_excel("file1.xlsx")

In [22]:
!cat file1.csv

,a,b,c
1,1,2,3
2,5,6,7
3,7,6,5


In [23]:
!cat file1.txt 

,a,b,c
1,1,2,3
2,5,6,7
3,7,6,5


In [25]:
!cat file1.json

{"a":{"1":1,"2":5,"3":7},"b":{"1":2,"2":6,"3":6},"c":{"1":3,"2":7,"3":5}}

In [0]:
# you can use this commands to download the files to your local system drive
from google.colab import files
files.download('file1.xlsx') 

##  Web Interaction
* It is possible to** interact** with **websites APIs** to retrieve 
**data** via a predefined **format**.

In [22]:
import requests
# url to get the first page of 30 issues of a GitHub Repository 
url1 = "https://api.github.com/repos/pandas-dev/pandas/issues"
# url to get the second page of 100 closed issues of a GitHub Repository, 
url2 = "https://api.github.com/repos/pandas-dev/pandas/issues?state=closed&page=2&per_page=100"


alliss= requests.get(url1)
closed= requests.get(url2)

# create DataFrames from the responses of the request
da= alliss.json()
daf=pd.DataFrame(da,columns=["id","state"])
# create DataFrames from 
dc= closed.json()
dcf=pd.DataFrame(dc,columns=["id","state"])

print(daf.shape)
print(dcf.shape)


(30, 2)
(100, 2)


In [15]:
pd.options.display.max_rows = 5

dcf

id   state
0   365425490  closed
1   365335732  closed
..        ...     ...
98  361451965  closed
99  361437710  closed

[100 rows x 2 columns]

In [18]:
daf

id state
0   370438271  open
1   370395722  open
..        ...   ...
28  369846876  open
29  369846349  open

[30 rows x 2 columns]

## Database Interaction


* In the following example, we will use **sqlachemy** and **pandas** to interact with an **sqlite** database.

* There is **various ways **to connect, create and extract data from a DataBase using sqlalchemy. We selected **one** of them.


In [2]:
# we need to install sqlalchemy
!pip install sqlalchemy 

In [25]:
import sqlalchemy as sqla
from sqlalchemy import Column, Table, types, MetaData

# if the database example1.db doesn't exist , the following statement will create it
DB= sqla.create_engine("sqlite:///A3P-w2-example1.db")

# to use "meta" later for the creation of the Table
meta = MetaData(DB)

# define a table's scheme
myTable=Table("myTab",meta,Column("id",types.String),Column("value",types.Integer))
# creation of the table
myTable.create()
# verification if the table exist
DB.has_table("myTab")



True

In [22]:

# verify if the corresponding file is created
!ls
# specify the  values to insert into the created table

insertion= myTable.insert().values([{"id": "a", "value":5},{"id": "b", "value":10},{"id": "c", "value":10}])
# insert the values nto the table
DB.execute(insertion)


A3P-w2-array_ex.txt		   A3P-w2-macrodata.csv
A3P-w2-csv_mindex.csv		   A3P-w2-out.csv
A3P-w2-ex1.csv			   A3P-w2-segismundo.txt
A3P-w2-ex1.xlsx			   A3P-w2-spx.csv
A3P-w2-ex2.csv			   A3P-w2-stinkbug.png
A3P-w2-ex3.csv			   A3P-w2-stock_px_2.csv
A3P-w2-ex3.txt			   A3P-w2-stock_px.csv
A3P-w2-ex4.csv			   A3P-w2-test_file.csv
A3P-w2-ex5.csv			   A3P-w2-tips.csv
A3P-w2-ex6.csv			   A3P-w2-tseries.csv
A3P-w2-ex7.csv			   A3P-w2-volume.csv
A3P-w2-example1.db		   A3P-w2-yahoo_price.pkl
A3P-w2-example.json		   A3P-w2-yahoo_volume.pkl
A3P-w2-fdic_failed_bank_list.html  README.md
A3P-w2-ipython_bug.py


In [23]:
# read the content of the table into a dataframe
pd.read_sql("myTab",DB)

id  value
0  a      5
1  b     10
2  c     10

In [10]:
!ls

A3P-w2-array_ex.txt		   A3P-w2-out.csv
A3P-w2-csv_mindex.csv		   A3P-w2-segismundo.txt
A3P-w2-ex1.csv			   A3P-w2-spx.csv
A3P-w2-ex1.xlsx			   A3P-w2-stinkbug.png
A3P-w2-ex2.csv			   A3P-w2-stock_px_2.csv
A3P-w2-ex3.csv			   A3P-w2-stock_px.csv
A3P-w2-ex3.txt			   A3P-w2-test_file.csv
A3P-w2-ex4.csv			   A3P-w2-tips.csv
A3P-w2-ex5.csv			   A3P-w2-tseries.csv
A3P-w2-ex6.csv			   A3P-w2-volume.csv
A3P-w2-ex7.csv			   A3P-w2-yahoo_price.pkl
A3P-w2-example.json		   A3P-w2-yahoo_volume.pkl
A3P-w2-fdic_failed_bank_list.html  example1.db
A3P-w2-ipython_bug.py		   README.md
A3P-w2-macrodata.csv


# 2- Missing Data

* Sometimes, data may have** missing** or **“Na”** values. So, with **pandas** we can **filter out** those values using the **dropna** method.

In [1]:
import numpy as np
from numpy import nan
from pandas import Series as S,  DataFrame as DF
# creating a Series with only "Na" values
ser1 = S(np.full(5,nan))
# affecting 7 and 3 to the first ([0]) second ([1])elements
ser1 [0:2]=[7,3]
ser1

0    7.0
1    3.0
2    NaN
3    NaN
4    NaN
dtype: float64

In [2]:
# drop the na values
ser1.dropna()

0    7.0
1    3.0
dtype: float64

In [3]:
df1=DF([ser1,ser1,ser1])
df1.iloc[0,[2,3]]=[9,10]
df1.iloc[1,3]=12
df1

0    1    2     3   4
0  7.0  3.0  9.0  10.0 NaN
1  7.0  3.0  NaN  12.0 NaN
2  7.0  3.0  NaN   NaN NaN

In [4]:
# drop the columns that have less than 2 observable values
df1.dropna(axis=1,thresh=2)

0    1     3
0  7.0  3.0  10.0
1  7.0  3.0  12.0
2  7.0  3.0   NaN

In [5]:
# drop the rows that have all values==nan
df1.dropna(how="all")

0    1    2     3   4
0  7.0  3.0  9.0  10.0 NaN
1  7.0  3.0  NaN  12.0 NaN
2  7.0  3.0  NaN   NaN NaN

* Instead  of **dropping** missing data, we can produce **new ones** using **pandas** with** fillna** method.

In [6]:
# replace no more than 2 nan values in each column by 0
df1.fillna(0,limit=2)

0    1    2     3    4
0  7.0  3.0  9.0  10.0  0.0
1  7.0  3.0  0.0  12.0  0.0
2  7.0  3.0  0.0   0.0  NaN

In [7]:
# by default axis =0 ==> will replace the first encoutred nan value (from top to bottom) with 
# the upper previous value
df1.fillna(method="ffill")

0    1    2     3   4
0  7.0  3.0  9.0  10.0 NaN
1  7.0  3.0  9.0  12.0 NaN
2  7.0  3.0  9.0  12.0 NaN

In [8]:
# axis =1 ==> will replace the first encoutred nan value (from left to right) with 
# the left previous value
df1.fillna(axis=1,method="ffill")

0    1    2     3     4
0  7.0  3.0  9.0  10.0  10.0
1  7.0  3.0  3.0  12.0  12.0
2  7.0  3.0  3.0   3.0   3.0

In [9]:
# df1 will be modified with inplace set to True
df1.fillna(0.99,inplace=True)
df1

0    1     2      3     4
0  7.0  3.0  9.00  10.00  0.99
1  7.0  3.0  0.99  12.00  0.99
2  7.0  3.0  0.99   0.99  0.99

# 3- Data Transformation

* Some other types of transforamtions are necessary as: **dropping duplicated ** data, **transforming** and creating new data using **mapping**, **renaming indexes**, **discretization**, **permutation** and **"random sampling"**



In [14]:
df1.drop_duplicates([0,1,2])

0    1     2     3     4
0  7.0  3.0  9.00  10.0  0.99
1  7.0  3.0  0.99  12.0  0.99

In [15]:
df1.drop_duplicates([0,1,2],keep='last')

0    1     2      3     4
0  7.0  3.0  9.00  10.00  0.99
2  7.0  3.0  0.99   0.99  0.99

In [16]:
# indexing a column that does not exist, will create it
df1["chars"]=list("abc")
df1

0    1     2      3     4 chars
0  7.0  3.0  9.00  10.00  0.99     a
1  7.0  3.0  0.99  12.00  0.99     b
2  7.0  3.0  0.99   0.99  0.99     c

In [17]:
# create a mapping using a dict
myMap ={ "a":"First letter","c":"Third letter","b":"Second letter"}
# create a new column using that mapping
df1["order"]=df1["chars"].map(myMap)
df1

0    1     2      3     4 chars          order
0  7.0  3.0  9.00  10.00  0.99     a   First letter
1  7.0  3.0  0.99  12.00  0.99     b  Second letter
2  7.0  3.0  0.99   0.99  0.99     c   Third letter

In [20]:
# transform ther order column values to uppercase
df1["order"]=df1["order"].str.upper()
df1

0    1     2      3     4 chars          order
0  7.0  3.0  9.00  10.00  0.99     a   FIRST LETTER
1  7.0  3.0  0.99  12.00  0.99     b  SECOND LETTER
2  7.0  3.0  0.99   0.99  0.99     c   THIRD LETTER

In [23]:
# replacing 0.99 values by 1 and  12 by 120
df2=df1.replace([0.99,12],[1,120])
df2


0    1    2      3    4 chars          order
0  7.0  3.0  9.0   10.0  1.0     a   FIRST LETTER
1  7.0  3.0  1.0  120.0  1.0     b  SECOND LETTER
2  7.0  3.0  1.0    1.0  1.0     c   THIRD LETTER

In [24]:
# replacing 0.99 values by 1 and  12 by 120 using a dictionary
df2=df1.replace({0.99:1,12:120})
df2

0    1    2      3    4 chars          order
0  7.0  3.0  9.0   10.0  1.0     a   FIRST LETTER
1  7.0  3.0  1.0  120.0  1.0     b  SECOND LETTER
2  7.0  3.0  1.0    1.0  1.0     c   THIRD LETTER

In [30]:
# renaming the indexes using a dict mapping
df2=df1.rename(index={0:"zero",1:"one",3:"three"})
df2

0    1     2      3     4 chars          order
zero  7.0  3.0  9.00  10.00  0.99     a   FIRST LETTER
one   7.0  3.0  0.99  12.00  0.99     b  SECOND LETTER
2     7.0  3.0  0.99   0.99  0.99     c   THIRD LETTER

In [39]:
ser2=S(np.arange(9))
ser2

0    0
1    1
2    2
3    3
4    4
5    5
6    6
7    7
8    8
dtype: int64

In [40]:
# grouping the ser2 values into four categories
import pandas as pd
pd.cut(ser2,[0,4,7,9])

0       NaN
1    (0, 4]
2    (0, 4]
3    (0, 4]
4    (0, 4]
5    (4, 7]
6    (4, 7]
7    (4, 7]
8    (7, 9]
dtype: category
Categories (3, interval[int64]): [(0, 4] < (4, 7] < (7, 9]]

In [41]:
# grouping the ser2 values into four categories with the same length
pd.cut(ser2,4)

0    (-0.008, 2.0]
1    (-0.008, 2.0]
2    (-0.008, 2.0]
3       (2.0, 4.0]
4       (2.0, 4.0]
5       (4.0, 6.0]
6       (4.0, 6.0]
7       (6.0, 8.0]
8       (6.0, 8.0]
dtype: category
Categories (4, interval[float64]): [(-0.008, 2.0] < (2.0, 4.0] < (4.0, 6.0] < (6.0, 8.0]]

In [46]:
# create an array with not ordred range values  (permuted)
norder= np.random.permutation(3)
norder

array([0, 2, 1])

In [48]:
# creating a new df reordred
df2.take(norder)

0    1     2      3     4 chars          order
zero  7.0  3.0  9.00  10.00  0.99     a   FIRST LETTER
2     7.0  3.0  0.99   0.99  0.99     c   THIRD LETTER
one   7.0  3.0  0.99  12.00  0.99     b  SECOND LETTER

In [53]:
# selecting randomly 5 values from ser2
ser2.sample(n=5)

0    0
6    6
7    7
3    3
1    1
dtype: int64

# 4- String Transfromation

* String object have useful methods that can be used:

In [57]:
# split a string specifying a separator
myStr= "This is an example"
splitted=myStr.split(" ")
splitted

['This', 'is', 'an', 'example']

In [58]:
# join strings with a separator
"-".join(splitted)

'This-is-an-example'

In [61]:
# replace a value in string by another value
newStr=myStr.replace(" ","_")
newStr

'This_is_an_example'

In [64]:
# find a value in a string using find
myStr.find("e")

11

In [65]:
# find a value in a string using index
myStr.index("e")

11

In [66]:
# find a value in a string using in
"e" in myStr

True

In [67]:
# the number of substring in a string
myStr.count(" ")

3

# References
* SQLAlchemy authors and contributors. Sqlalchemy 1.2 documentation. On-line at https://docs.sqlalchemy.org/en/latest/core/dml.html. Ac-cessed on 19-10-2018.




* GitHub. Rest api v3. On-line at https://developer.github.com/v3/.
Accessed on 15-10-2018.
* Wes McKinney. Python for data analysis: Data wrangling with Pandas, NumPy, and IPython. O’Reilly Media, Inc, 2018.
* pydata.org. Pandas documentation. On-line at https://pandas.pydata.org/. Accessed on 19-10-2018.
* pysheeet. Python sqlalchemy cheatsheet. On-line at https://pysheeet.readthedocs.io/en/latest/notes/python-sqlalchemy.html. Accessedon 19-10-2018.
* McKinney Wes. pydata-book. On-line at https://github.com/wesm/pydata-book.git. Accessed on 14-10-2018